analytics/src/main/java/com/segment/analytics/IntegrationOperation.java

```diff
@@ -154,141 +159,159 @@ public String toString() {
     };
   }
 
-  static IntegrationOperation identify(final IdentifyPayload identifyPayload) {
-    return new IntegrationOperation() {
-      @Override
-      public void run(String key, Integration<?> integration, ProjectSettings projectSettings) {
-        if (isIntegrationEnabled(identifyPayload.integrations(), key)) {
-          integration.identify(identifyPayload);
-        }
-      }
+  @Private
+  static List<Middleware> getMiddlewareList(
+      Map<String, List<Middleware>> destinationMiddleware, String key) {
+    List<Middleware> middleware = destinationMiddleware.get(key);
+    if (middleware == null) {
+      // No middleware registered, return empty list
+      return Collections.emptyList();
+    }
+    return middleware;
+  }
 
-      @Override
-      public String toString() {
-        return identifyPayload.toString();
-      }
-    };
+  @Private
+  static void runMiddlewareChain(
+      BasePayload payload, List<Middleware> middleware, Middleware.Callback callback) {
+    Middleware.Chain chain = new MiddlewareChainRunner(0, payload, middleware, callback);
+    chain.proceed(payload);
   }
 
-  static IntegrationOperation group(final GroupPayload groupPayload) {
+  /**
+   * Integration Operation for a Segment Event (track | identify | alias | group | screen) Operation
+   * runs destination middleware for given integration before sending to the desired integration
+   */
+  static IntegrationOperation segmentEvent(
+      final BasePayload payload, Map<String, List<Middleware>> destinationMiddleware) {
     return new IntegrationOperation() {
       @Override
-      public void run(String key, Integration<?> integration, ProjectSettings projectSettings) {
-        if (isIntegrationEnabled(groupPayload.integrations(), key)) {
-          integration.group(groupPayload);
-        }
+      void run(String key, Integration<?> integration, ProjectSettings projectSettings) {
+        List<Middleware> applicableMiddleware = getMiddlewareList(destinationMiddleware, key);
+        runMiddlewareChain(
+            payload,
+            applicableMiddleware,
```

Comment:
the isIntegrationEnabled bit that seems to have been removed here seems like it might be important.  If an integration is disabled on segment.com, there's no need to run it's middleware right?

Reply:
never mind, I see it down below.  however, it's repeated.  seems like that could be placed here instead and checked once?